includes/multi-currency/MultiCurrency.php

```diff
@@ -1184,27 +1154,106 @@ private function register_admin_scripts() {
 		);
 	}
 
+	/**
+	 * Validates the given currency code.
+	 *
+	 * @param   string $currency_code  The currency code to check validity.
+	 *
+	 * @return  string|false  Returns back the currency code in uppercase letters if it's valid, or `false` if not.
+	 */
+	public function validate_currency_code( $currency_code ) {
+		return in_array( strtoupper( $currency_code ), array_keys( $this->available_currencies ), true )
+		? strtoupper( $currency_code )
+		: false;
+	}
+
+	/**
+	 * Get simulation params from querystring and activate when needed
+	 *
+	 * @return  void
+	 */
+	public function possible_simulation_activation() {
+		// This is required in the MC onboarding simulation iframe.
+		$this->simulation_params = $this->get_multi_currency_onboarding_simulation_variables();
+		if ( ! $this->is_simulation_enabled() ) {
+			return;
+		}
+		// Modify the page links to deliver required params in the simulation.
+		$this->add_simulation_params_to_preview_urls();
+		$this->simulate_client_currency();
+	}
+
+	/**
+	 * Enables simulation of client browser currency.
+	 *
+	 * @return  void
+	 */
+	public function simulate_client_currency() {
+		if ( ! $this->simulation_params['enable_auto_currency'] || ! $this->simulation_params['client_currency'] ) {
+			return;
+		}
+		$simulation_currency = $this->simulation_params['client_currency'];
+
+		// Simulate client currency from geolocation.
+		add_filter(
+			'wcpay_multi_currency_get_selected_currency',
+			function( $selected_currency ) use ( $simulation_currency ) {
+				return $simulation_currency;
+			}
+		);
+
+		// Always display the notice on simulation screen.
+		add_action( 'wp_footer', [ $this, 'display_geolocation_currency_update_notice' ] );
+
+		// Skip recalculating the cart to prevent infinite loop in simulation.
+		remove_action( 'wp_loaded', [ $this, 'recalculate_cart' ] );
+
+	}
+
+	/**
+	 * Returns whether the simulation querystring param is set and active
+	 *
+	 * @return  bool  Whether the simulation is enabled or not
+	 */
+	public function is_simulation_enabled() {
+		return 0 < count( array_keys( $this->simulation_params ) );
+	}
+
 	/**
 	 * Gets the multi currency onboarding preview overrides from the querystring.
 	 *
 	 * @return  array  Override variables
 	 */
 	public function get_multi_currency_onboarding_simulation_variables() {
```

Comment:
I suggest adding unit tests for this method, since the logic is not trivial

Reply:
How about converting it to `private`?

Reply:
That works, would be great to have some high-level test though that covers that private method 👍

Reply:
Fixed.